src/xkbcomp/rules.c

```diff
@@ -1101,13 +1101,37 @@ read_rules_file(struct xkb_context *ctx,
 
     /* Skip UTF-8 encoded BOM (U+FEFF) */
     /* See: https://www.unicode.org/faq/utf_bom.html#bom5 */
-    scanner_str(&scanner, "\xef\xbb\xbf", 3);
+    if (scanner_str(&scanner, "\xef\xbb\xbf", 3) || scanner.len < 2) {
+        goto initial;
+    }
 
-    ret = matcher_match(matcher, &scanner, include_depth, string, size, path);
+    /* Early detection of wrong file encoding, e.g. UTF-16 or UTF-32 */
+    if (scanner.s[0] == '\0' || scanner.s[1] == '\0') {
+        if (scanner.s[0] != '\0')
+            scanner.token_column++;
+        scanner_err(&scanner, "unexpected NULL character.");
+        goto encoding_error;
+    }
+    if (!is_ascii(scanner.s[0])) {
+        scanner_err(&scanner, "unexpected non-ASCII character.");
+        goto encoding_error;
+    }
```

Comment:
tbh, this code is bit confusing with the interleaved gotos. the `out` label is not (no longer?) needed, we can return from the error condition after map_file.

but this here should be a helper function so the code reads as something like
```c
if (is_supported_encoding()) {
   scanner_err(&scanner, "yuck!");
   return false;
}

ret = matcher_match(...);
```

that should make it possible to skip all the gotos, or most of them. I think the same applies to the other patches with similar approaches.

Note there's also a bug afaict, skipping to `encoding_error` never calls `unmap_file()` in 1122.

Reply:
Good advice! That’s why I marked the PR as draft.

Refactored. I chose `check_supported_char_encoding` over `is_supported_encoding` because of the BOM detection, which in fact _consumes_ the input.

I also improved the comments.

Reply:
thanks, looking much better now, a bit of deduplication [^1] would be nice though if possible (see my comment below)

[^1]: detriplication, to be technical :)